File: types/api/src/api/worksheet/print.ts
/**
 * WorksheetPrint — Sub-API Interface for Print Settings and Page Breaks
 *
 * Methods for managing print settings, print area, and manual page breaks.
 */
import type { HeaderFooterImageInfo, HfImagePosition, PageMargins } from '@mog/types-core';
import type { PrintSettings } from '../types';

/** Sub-API for worksheet print and page break operations. */
export interface WorksheetPrint {
  /**
   * Get the current print settings for the sheet.
   *
   * @returns Current print settings
   */
  getSettings(): Promise<PrintSettings>;

  /**
   * Update print settings. Only the provided keys are changed.
   *
   * @param settings - Partial print settings to apply
   */
  setSettings(settings: Partial<PrintSettings>): Promise<void>;

  /**
   * Get the current print area as an A1-notation range string.
   *
   * @returns The print area range string, or null if no print area is set
   */
  getArea(): Promise<string | null>;

  /**
   * Set the print area to the specified A1-notation range.
   *
   * @param area - A1-notation range string (e.g., "A1:H20")
   */
  setArea(area: string): Promise<void>;

  /**
   * Clear the print area so the entire sheet prints.
   */
  clearArea(): Promise<void>;

  /**
   * Add a manual page break at the specified position.
   *
   * @param type - 'horizontal' (row break) or 'vertical' (column break)
   * @param position - 0-based row or column index for the break
   */
  addPageBreak(type: 'horizontal' | 'vertical', position: number): Promise<void>;

  /**
   * Remove a manual page break at the specified position.
   *
   * @param type - 'horizontal' or 'vertical'
   * @param position - 0-based row or column index of the break to remove
   */
  removePageBreak(type: 'horizontal' | 'vertical', position: number): Promise<void>;

  /**
   * Get all manual page breaks in the sheet.
   *
   * @returns Object containing arrays of horizontal and vertical break positions
   */
  getPageBreaks(): Promise<{
    rowBreaks: Array<{ id: number; min: number; max: number; manual: boolean; pt: boolean }>;
    colBreaks: Array<{ id: number; min: number; max: number; manual: boolean; pt: boolean }>;
  }>;

  /**
   * Remove all manual page breaks from the sheet.
   */
  clearPageBreaks(): Promise<void>;

  /**
   * Set the rows to repeat at the top of each printed page (print titles).
   *
   * @param startRow - 0-based start row index
   * @param endRow - 0-based end row index (inclusive)
   */
  setPrintTitleRows(startRow: number, endRow: number): Promise<void>;

  /**
   * Set the columns to repeat at the left of each printed page (print titles).
   *
   * @param startCol - 0-based start column index
   * @param endCol - 0-based end column index (inclusive)
   */
  setPrintTitleColumns(startCol: number, endCol: number): Promise<void>;

  /**
   * Get the rows configured to repeat at the top of each printed page.
   *
   * @returns A [startRow, endRow] tuple (0-based, inclusive), or null if no repeat rows are set
   */
  getPrintTitleRows(): Promise<[number, number] | null>;

  /**
   * Get the columns configured to repeat at the left of each printed page.
   *
   * @returns A [startCol, endCol] tuple (0-based, inclusive), or null if no repeat columns are set
   */
  getPrintTitleColumns(): Promise<[number, number] | null>;

  /**
   * Clear all print titles (both repeat rows and repeat columns).
   * Other print settings (margins, orientation, etc.) are preserved.
   */
  clearPrintTitles(): Promise<void>;

  /**
   * Set page margins with unit conversion.
   * Values are converted to inches (OOXML native unit) before storing.
   * Only provided margin fields are updated; others are preserved.
   *
   * @param unit - Unit of the provided values: 'inches', 'points', or 'centimeters'
   * @param options - Partial margin values to set
   */
  setPrintMargins(
    unit: 'inches' | 'points' | 'centimeters',
    options: Partial<PageMargins>,
  ): Promise<void>;

  /**
   * Get the cell position immediately after a page break.
   *
   * @param type - Break type: 'horizontal' (row break) or 'vertical' (column break)
   * @param position - 0-based row or column index of the break
   * @returns Cell coordinates of the first cell after the break
   */
  getCellAfterBreak(
    type: 'horizontal' | 'vertical',
    position: number,
  ): { row: number; col: number };

  /**
   * Get all header/footer images for this sheet.
   *
   * @returns Array of image info objects, one per occupied position
   */
  getHeaderFooterImages(): Promise<HeaderFooterImageInfo[]>;

  /**
   * Set or replace a header/footer image at the specified position.
   * The engine automatically manages the `&G` format code in the corresponding
   * header/footer string.
   *
   * @param info - Image info including position, src (data-URL or path), and dimensions
   */
  setHeaderFooterImage(info: HeaderFooterImageInfo): Promise<void>;

  /**
   * Remove the header/footer image at the specified position.
   * The engine automatically removes the `&G` format code from the corresponding
   * header/footer string.
   *
   * @param position - Which of the 6 header/footer positions to clear
   */
  removeHeaderFooterImage(position: HfImagePosition): Promise<void>;
}

File: compute/api/src/sheet/print.rs
//! SheetPrint — Print settings, page breaks, print area, and print titles.

use crate::dispatch::Dispatch;
use crate::error::ComputeApiError;
use cell_types::SheetId;
use domain_types::domain::print::{PageBreaks, PrintSettings};
use domain_types::{PrintRange, PrintTitles};
use snapshot_types::MutationResult;

/// Print-related operations for a single sheet.
///
/// Manages page breaks (horizontal/vertical), print area, print titles
/// (repeating rows/columns), and general print settings (orientation,
/// margins, scaling, etc.).
pub struct SheetPrint {
    dispatch: Dispatch,
    sheet_id: SheetId,
}

impl SheetPrint {
    pub(crate) fn new(dispatch: Dispatch, sheet_id: SheetId) -> Self {
        Self { dispatch, sheet_id }
    }

    // -----------------------------------------------------------------
    // Page breaks
    // -----------------------------------------------------------------

    /// Get all page breaks for the sheet.
    pub fn get_page_breaks(&self) -> Result<PageBreaks, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch.query_engine(move |e| e.get_page_breaks(&sid))
    }

    /// Add a horizontal page break before the given row.
    pub fn add_horizontal_page_break(&self, row: u32) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.add_horizontal_page_break(&sid, row).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    /// Remove a horizontal page break at the given row.
    pub fn remove_horizontal_page_break(
        &self,
        row: u32,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.remove_horizontal_page_break(&sid, row).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    /// Add a vertical page break before the given column.
    pub fn add_vertical_page_break(&self, col: u32) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.add_vertical_page_break(&sid, col).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    /// Remove a vertical page break at the given column.
    pub fn remove_vertical_page_break(&self, col: u32) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.remove_vertical_page_break(&sid, col).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    /// Clear all page breaks for the sheet.
    pub fn clear_all_page_breaks(&self) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.clear_all_page_breaks(&sid).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    // -----------------------------------------------------------------
    // Print area
    // -----------------------------------------------------------------

    /// Get the print area for the sheet.
    pub fn get_print_area(&self) -> Result<Option<PrintRange>, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch.query_engine(move |e| e.get_print_area(&sid))
    }

    /// Set or clear the print area for the sheet.
    pub fn set_print_area(
        &self,
        area: Option<PrintRange>,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.set_print_area(&sid, area).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    // -----------------------------------------------------------------
    // Print titles
    // -----------------------------------------------------------------

    /// Get print titles (repeating rows/columns) for the sheet.
    pub fn get_print_titles(&self) -> Result<PrintTitles, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .query_engine(move |e| e.get_print_titles(&sid))
    }

    /// Set print titles (repeating rows/columns) for the sheet.
    pub fn set_print_titles(&self, titles: PrintTitles) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.set_print_titles(&sid, titles).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    // -----------------------------------------------------------------
    // Print settings
    // -----------------------------------------------------------------

    /// Set print settings (orientation, margins, scaling, etc.) for the sheet.
    pub fn set_print_settings(
        &self,
        settings: PrintSettings,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.set_print_settings(&sid, settings).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    // -----------------------------------------------------------------
    // Header/footer images
    // -----------------------------------------------------------------

    /// Get all header/footer images for the sheet.
    pub fn get_hf_images(
        &self,
    ) -> Result<Vec<domain_types::domain::print::HeaderFooterImageInfo>, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch.query_engine(move |e| e.get_hf_images(&sid))
    }

    /// Set or replace a header/footer image at the specified position.
    pub fn set_hf_image(
        &self,
        info: domain_types::domain::print::HeaderFooterImageInfo,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.set_hf_image(&sid, info).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }

    /// Remove the header/footer image at the specified position.
    pub fn remove_hf_image(
        &self,
        position: domain_types::domain::print::HfImagePosition,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.remove_hf_image(&sid, position).map(|(_, r)| r))
            .and_then(|r| r.map_err(ComputeApiError::from))
    }
}
